lib/class-wp-theme-json-gutenberg.php

```diff
@@ -664,22 +664,28 @@ private static function compute_preset_classes( $settings, $selector ) {
 
 		$stylesheet = '';
 		foreach ( self::PRESETS_METADATA as $preset ) {
-			$values = _wp_array_get( $settings, $preset['path'], array() );
-			foreach ( $values as $value ) {
-				foreach ( $preset['classes'] as $class ) {
-					$stylesheet .= self::to_ruleset(
-						// We don't want to use kebabCase here,
-						// see https://github.com/WordPress/gutenberg/issues/32347
-						// However, we need to make sure the generated class
-						// doesn't contain spaces.
-						self::append_to_selector( $selector, '.has-' . preg_replace( '/\s+/', '-', $value['slug'] ) . '-' . $class['class_suffix'] ),
-						array(
+			$values_per_origin = _wp_array_get( $settings, $preset['path'], array() );
```

Comment:
If I understood properly this [PR's direction](https://github.com/WordPress/gutenberg/pull/32358#issuecomment-853282220), it seems that we'll always enqueue all the preset values (classes + custom properties) for every origin.

For example, if both core and a theme provides a color with `yellow` slug, the classes and custom properties for them will be enqueued in the order we process the origins, so the later wins. If then user modifies the value of that yellow we end up with:

```css
body {
  --wp--preset--color--yellow: <core_value>;
  --wp--preset--color--yellow: <theme_value>;
  --wp--preset--color--yellow: <user_value>;
}

.has-yellow-color { <core_value> !important; };
.has-yellow-color { <theme_value> !important; };
.has-yellow-color { <user_value> !important; };
```

I'm wondering if a different implementation for `compute_preset_classes` and `compute_preset_vars` would make the code clearer and remove the duplicates. What if we:

1. Reduced the `settings.color.palette` to a single object with slug and value. This step already takes care of the duplicates (the later origin to be processed wins). Something like:

```js
{
  <slug-1>: <value-1>,
  <slug-2>: <value-2>,
  ...
}
```

2. Maped this object to the functions's output (either classes or custom properties).

Reply:
Actually. More thoughts for when two yellows are defined in different origins. Overwriting the previous origin works for now with the use case we have in mind: make those classes and custom properties available to patterns, independently from the theme in use.

However, when we want to show the colors in separate slots in the UI, how does that work if they have the same slug? Both when themes provide a color with the same slug than core's and when the user edits the value of a core or theme color. (Not that we need to _implement_ it in this PR, but, at least, we should _understand_ how feasible it is using this approach.)

Reply:
I applied your feedback! In the future, when implementing the UI, we may need to do something smart, and if the a color with the same slug existis in a higher priority origin, the UI stops showing the slug with lowest priority

Reply:
A braindump of the options I see:

1. If theme colors use an existing core slug, the core color is overwritten. The current proposal. Gives us a way to update core colors, we'd need something separate if we wanted to remove them (perhaps a `coreColors: false` or a list of allowed colors).
2. If theme colors use an existing core slug, the theme color is removed. The way themes update/remove core colors is by updating a specific key in `theme.json`, such as `coreColors: [ ... ]`.

Given that the reason we want core colors available is to have consistent CSS for use in patterns, removing them without alternative doesn't seem like something we need. So, it boils down to optimize for how we want updating to work. The option 2 is more intentional and straightforward (doesn't require we do anything _smart_ for the UI).

If we went with option 2, we could do it in a follow up.

Reply:
I think what we have right now is enough we can overwrite the previous color slugs. There is now a way to remove them and that's our objective, in previous versions of WordPress there is also no way to remove core color classes one can just overwrite, we are doing the same behavior.